### ISSUE TITLE
pyright-lsp cannot find some local library when the current directory has same name as some pip library.

### ISSUE THREAD

**Issue by zhou13** [microsoft/pyright#5067]
**Describe the bug**
After #4994 been fixed, I still have some trouble using pyright-lsp on my project. Here is the issue.

**To Reproduce**

1. Download [bugreproduce.zip](https://github.com/microsoft/pyright/files/11389656/bugreproduce.zip)

2. Uncompress it to /private/tmp. The structure should look like
```
$ tree /private/tmp/bugreproduce
├── [  16]  logname-arguments
├── [3.2K]  logname-env
├── [9.3K]  logname-stdin
├── [5.5K]  logname-stdout
├── [  31]  pyproject.toml
├── [   0]  requirements.txt
├── [ 420]  setup.py
└── [ 160]  src
    ├── [  14]  foobar.py
    └── [ 192]  projects
        ├── [  54]  example.py
        ├── [ 226]  requirements.txt
        └── [ 160]  sac
            ├── [   0]  __init__.py
            └── [  21]  datasets.py
```
3. Install self using pip in a virtualenv
```
cd /private/tmp/bugreproduce
conda create -n bugreproduce python=3.8
source activate bugreproduce
pip install -e .
```
4. replay the problem using pyright 1.1.306. Older version don't work.
```
awk '{print $0; system("sleep .3");}' logname-stdin | pyright-langserver --stdio
```

I got
```
$ awk '{print $0; system("sleep .3");}' logname-stdin | ~/.local/share/nvim/mason/packages/pyright/node_modules/.bin/pyright-langserver --stdio
Content-Length: 119

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Pyright language server 1.1.306 starting"}}Content-Length: 187

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Server root directory: /Users/yichaozhou/.local/share/nvim/mason/packages/pyright/node_modules/pyright/dist/"}}Content-Length: 147

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Starting service instance \"/private/tmp/bugreproduce/src/projects\""}}Content-Length: 1046

{"jsonrpc":"2.0","id":1,"result":{"capabilities":{"textDocumentSync":2,"definitionProvider":{"workDoneProgress":true},"declarationProvider":{"workDoneProgress":true},"typeDefinitionProvider":{"workDoneProgress":true},"referencesProvider":{"workDoneProgress":true},"documentSymbolProvider":{"workDoneProgress":true},"workspaceSymbolProvider":{"workDoneProgress":true},"hoverProvider":{"workDoneProgress":true},"documentHighlightProvider":{"workDoneProgress":true},"renameProvider":{"prepareProvider":true,"workDoneProgress":true},"completionProvider":{"triggerCharacters":[".","[","\"","'"],"resolveProvider":true,"workDoneProgress":true,"completionItem":{"labelDetailsSupport":true}},"signatureHelpProvider":{"triggerCharacters":["(",",",")"],"workDoneProgress":true},"codeActionProvider":{"codeActionKinds":["quickfix","source.organizeImports"],"workDoneProgress":true},"executeCommandProvider":{"commands":[],"workDoneProgress":true},"callHierarchyProvider":true,"workspace":{"workspaceFolders":{"supported":true,"changeNotifications":true}}}}}Content-Length: 160

{"jsonrpc":"2.0","id":0,"method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///private/tmp/bugreproduce/src/projects","section":"python"}]}}Content-Length: 160

{"jsonrpc":"2.0","id":1,"method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///private/tmp/bugreproduce/src/projects","section":"python"}]}}Content-Length: 169

{"jsonrpc":"2.0","id":2,"method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///private/tmp/bugreproduce/src/projects","section":"python.analysis"}]}}Content-Length: 169

{"jsonrpc":"2.0","id":3,"method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///private/tmp/bugreproduce/src/projects","section":"python.analysis"}]}}Content-Length: 161

{"jsonrpc":"2.0","id":4,"method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///private/tmp/bugreproduce/src/projects","section":"pyright"}]}}Content-Length: 161

{"jsonrpc":"2.0","id":5,"method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///private/tmp/bugreproduce/src/projects","section":"pyright"}]}}Content-Length: 99

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Found 3 source files"}}Content-Length: 99

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Found 3 source files"}}Content-Length: 36

{"jsonrpc":"2.0","id":2,"result":[]}Content-Length: 165

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///private/tmp/bugreproduce/src/projects/example.py","version":0,"diagnostics":[]}}Content-Length: 36

{"jsonrpc":"2.0","id":3,"result":[]}Content-Length: 134

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Starting service instance \"/private/tmp/bugreproduce\""}}Content-Length: 147

{"jsonrpc":"2.0","id":6,"method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///private/tmp/bugreproduce","section":"python"}]}}Content-Length: 156

{"jsonrpc":"2.0","id":7,"method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///private/tmp/bugreproduce","section":"python.analysis"}]}}Content-Length: 148

{"jsonrpc":"2.0","id":8,"method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///private/tmp/bugreproduce","section":"pyright"}]}}Content-Length: 150

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Loading pyproject.toml file at /private/tmp/bugreproduce/pyproject.toml"}}Content-Length: 177

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":1,"message":"Pyproject file \"/private/tmp/bugreproduce/pyproject.toml\" is missing \"[tool.pyright]\" section."}}Content-Length: 99

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Found 5 source files"}}Content-Length: 36

{"jsonrpc":"2.0","id":4,"result":[]}Content-Length: 36

{"jsonrpc":"2.0","id":5,"result":[]}Content-Length: 155

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///private/tmp/bugreproduce/src/foobar.py","version":0,"diagnostics":[]}}Content-Length: 480

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///private/tmp/bugreproduce/src/projects/example.py","version":4,"diagnostics":[{"range":{"start":{"line":4,"character":5},"end":{"line":4,"character":17}},"message":"Import \"sac.datasets\" could not be resolved","severity":2,"code":"reportMissingImports","source":"Pyright","codeDescription":{"href":"https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportMissingImports"}}]}}Content-Length: 165

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///private/tmp/bugreproduce/src/projects/example.py","version":4,"diagnostics":[]}}Content-Length: 36

{"jsonrpc":"2.0","id":6,"result":[]}Content-Length: 165

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///private/tmp/bugreproduce/src/projects/example.py","version":5,"diagnostics":[]}}Content-Length: 480

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///private/tmp/bugreproduce/src/projects/example.py","version":5,"diagnostics":[{"range":{"start":{"line":5,"character":5},"end":{"line":5,"character":17}},"message":"Import \"sac.datasets\" could not be resolved","severity":2,"code":"reportMissingImports","source":"Pyright","codeDescription":{"href":"https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportMissingImports"}}]}}Content-Length: 36
```

It says `sac.datasets` cannot be resolved. But it is a local directory. Note that step 3 is necessary for reproducing this bug.

**Expected behavior**

`sac.datasets` should be resolvable.

**Comment by erictraut** [comment]
I suspect this is a user configuration issue. If you haven't already done so, please review the [import resolution documentation](https://microsoft.github.io/pyright/#/import-resolution) and try to resolve the problem yourself. If you still think that it's a possible bug, I'll look at it.

**Comment by zhou13** [comment]
https://user-images.githubusercontent.com/576382/236015498-6185d088-d9ab-43de-9f82-271f0f68bfec.mov

I attached a video to show the behavior I saw. I think it is a bug since the language server works well before I open a second workspace.

**Comment by erictraut** [comment]
The import resolution paths for each workspace will differ because the root of the workspace differs. If the same file is found in two workspaces, import resolution may succeed in one workspace but not in the other.

Why are you opening multiple overlapping workspaces? This is not typical, and you will likely hit a number of issues, some of them "as designed". I recommend opening a single workspace that has as its root directory the root of your project. If you want to use multiple workspaces, make sure they are not overlapping.

**Comment by erictraut** [comment]
You can diagnose import resolution issues by setting "verboseOutput" to true in the configuration file. This will cause pyright to log detailed information for each import resolution failures.

**Comment by zhou13** [comment]
> Why are you opening multiple overlapping workspaces?

We have a main library and several libraries that use the main library in the same repo. The intuitive structure is
```
pyproject.yaml
src/
  main_library/
    ....
  projects/
     main_library_user1/
         ....
     main_library_user2/
```
When doing debugging and development, it is common to jump from `main_library_user1` to a file in `main_library` and thus will trigger the issue.

I put the `verboseOutput` into `/private/tmp/bugreproduce/pyrightconfig.json` and `/private/tmp/bugreproduce/src/projects/pyrightconfig.json` but it seems pyright does not give additional information when running `awk '{print $0; system("sleep .3");}' logname-stdin | pyright-langserver --stdio`.

```
$ cat pyrightconfig.json
{
  "verboseOutput": true
}
$ cat src/projects/pyrightconfig.json
{
  "verboseOutput": true
}
```

**Comment by zhou13** [comment]
You are right that it seems to be a resolution issue. If I remove `src/projects/requirements.txt` then it will show the problem on the first opening.

**Comment by erictraut** [comment]
The problem is that you have overlapping workspace roots, so some files (like `example.py`) are part of multiple workspaces. Since each workspace has its own settings (including import resolution paths), files within each workspace are analyzed within the context of its workspace. Files like `example.py` are analyzed multiple times, once for each workspace they belong to. When the diagnostic list is pushed to the client, there's no mechanism to report "here are the diagnostics for this file associated with this workspace root". The LSP "publishDiagnostic" message contains only a "uri" field, not a "workspaceRoot" field. That means pyright reports diagnostics multiple times for this file, and the last reported diagnostics "win". It's therefore based on timing, which is nondeterministic.

I think pyright is working correctly here, so I'm going to close this.

I recommend that you don't configure workspace roots that overlap. If you must have overlapping workspace directories, then you should create pyright configurations (either `pyrightconfig.json` or `pyproject.toml`) in each of the workspace roots that specify which subdirectories to "include" and "exclude" for that workspace.

**Comment by zhou13** [comment]
> I recommend that you don't configure workspace roots that overlap. If you must have overlapping workspace directories, then you should create pyright configurations (either pyrightconfig.json or pyproject.toml) in each of the workspace roots that specify which subdirectories to "include" and "exclude" for that workspace.

Thanks for the suggestion and help! @erictraut Sorry for commenting on closed issues. I am reading [configuration document](https://github.com/microsoft/pyright/blob/main/docs/configuration.md) of `exclude`. Just wonder whether you think it is a good idea to exclude sub-directories that are recognized as a workspace by default. It should reduce the number of cases where one file is part of multiple workspaces, which is not what people normally want and might generate hard-to-debug problems like this one.
